### PR TITLE
c++ optimizations

### PR DESCRIPTION
**Pull request by d-a-v**
Thank you very much for this cpu-only very efficient and portable implementation.
Here's my two cents. Same algorithm, same results, faster.

### CODE REVIEW

File: BarnzNhutt.cpp [hunk 1]
```diff
@@ -235,17 +238,14 @@ void createFrame(char* image, double* hdImage, struct body* b, int step)
...
-		hdImage[i] = 0.0;
-	}
+	memset(image, 0, WIDTH*HEIGHT*3);
+	memset(hdImage, 0, WIDTH*HEIGHT*3*sizeof(double));
 }
 
 void renderBodies(struct body* b, double* hdImage)
 {
 	/// ORTHOGONAL PROJECTION
+	#pragma omp parallel
```
Comment (PWhiddy) [inline BarnzNhutt.cpp]: This seems to be slower than the serial version- at least with the system I'm currently testing on

**Review by PWhiddy — COMMENTED** [review]
Thanks for these great optimizations!
A couple note upon first testing:
There is a " 'eenableexcept' was not declared in this scope" error when compiling on macos 10.13 with gcc-9
Removing this line I was able to build fine. Maybe this is not an issue on linux.
I saw over a 2x speedup in the simulation code! Amazing work.
However, I actually noticed a significant slowdown in the rendering code. 
I think that this is because the inner loop is too small to benefit from being parallel. 
(As a side note, I do think its worth mentioning here that even though there is technically a race condition when rendering all particles in parallel, unless all the particles are occupying the same small area of the screen the effects of this race condition are probably very minimal. It might be okay to let different threads overwrite each other in rare cases)

I definitely want to merge these changes to the simulation code, but will wait to see if you have thoughts on these couple issues mentioned above.

Going to do some more testing tomorrow on an ec2 instance to see how it handles there. 
Thanks again!

**Comment by d-a-v** [comment]
Parallelizing rendering is indeed problematic, because of the possible race as you mentioned, and the poor efficiency when parallelizing single-dot-rendering. I decoupled `parallel` and `for` so the threads are already started and noticed a small improvement but maybe not worth leaving it as it is not efficient in every cases.
I will remove this parallelization from my PR, and make the FPE always optional (it helps when debugging but can break a long-run simulation), and also make it only enabled on linux unless I can find and test the equivalent on macOS.

Another optimization path:
Pipe `ffmpeg` input directly from your code so we don't get through intermediate files that also take ages or high price to be written on a filesystem.

Further thoughts:
In another branch I am also going to decouple rendering from processing, and try if I can use [Threejs particles](https://threejs.org/examples/#webgl_gpu_particle_system) to enable live rendering, live exploring and possibly live interaction. To talk with javascript on browser websockets are needed. I'm not a js specialist but I can try to setup an environment that would allow others to improve it. 
Another path is to pipe `ffmpeg` directly from your code so we don't get through intermediate files that also take ages or high price to be written on a filesystem.
(*edit*: [pipe to ffmpeg](https://stackoverflow.com/questions/51903888/is-it-possible-to-send-ffmpeg-images-by-using-pipe))

Thank you again for your implementation. I tried about tens of them before finding yours. Efficient, portable, and surely one of the best when no GPU are available or too proprietary.

**Comment by PWhiddy** [comment]
Looks good! Thanks again, I'm glad its useful for you.

Regarding your idea to render the particles with threejs, that does sound like a cool idea. I have actually implemented an interactive particle system which runs the simulation on the backend, and uses sockets to transfer particle data and user interaction to/from the client. I was using node for the backend. 

Source:
https://github.com/PWhiddy/OurGalaxy

A hosted demo:
https://floating-bayou-46849.herokuapp.com/ (click and drag to add particles)

This was just a minimal proof of concept, but may be a useful reference. I would be very interested in the further development of this idea. Keep me posted!

**Pull request merged**